### PR TITLE
Sparklines implementation

### PR DESCRIPTION
**Pull request by Pankraty**
I finally managed to get back to this task and now it is ready for review.

This PR replaces #907 and implements #454.

cc @ghronkrepps

### CODE REVIEW

**Comment by igitur** [comment]
Please rebase.

**Comment by igitur** [comment]
Please rebase and I will look at it soon after.

**Comment by Pankraty** [comment]
Yay, I did it! 😺

File: ClosedXML/Excel/XLWorksheet.cs [hunk 1]
```diff
@@ -1301,7 +1302,28 @@ private void ShiftDataValidationColumns(XLRange range, int columnsShifted)
         // TODO: Sparklines
         private void ShiftSparklineColumns(XLRange range, int columnsShifted)
         {
-            throw new NotImplementedException();
+            var model = new XLRangeAddress(
```
Comment (igitur) [inline ClosedXML/Excel/XLWorksheet.cs]: I think you can remove the `TODO` here.

File: ClosedXML/Excel/XLWorksheet.cs [hunk 2]
```diff
@@ -1434,7 +1457,28 @@ private void ShiftDataValidationRows(XLRange range, int rowsShifted)
         // TODO: Sparklines
         private void ShiftSparklineRows(XLRange range, int rowsShifted)
         {
-            throw new NotImplementedException();
+            var model = new XLRangeAddress(
```
Comment (igitur) [inline ClosedXML/Excel/XLWorksheet.cs]: I think you can remove the `TODO` here.

File: ClosedXML/Excel/Cells/XLCellCopyOptions.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
```
Comment (igitur) [inline ClosedXML/Excel/Cells/XLCellCopyOptions.cs]: I think the introduction of `XLCellCopyOptions` should be its own commit, preferrably the first one, before sparklines are added.

File: ClosedXML/Excel/Ranges/XLRangeBase.cs [hunk 4]
```diff
@@ -1657,9 +1665,10 @@ public void Delete(XLShiftDeletedCells shiftDeleteCells)
 
                     if (canInsert)
                         cellsToInsert.Add(newCellAddress, c);
-                }
+            }
```
Comment (igitur) [inline ClosedXML/Excel/Ranges/XLRangeBase.cs]: Indentation problem here.

File: ClosedXML_Examples/Sparklines/SampleSparklines.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            ws3.SparklineGroups.ForEach(g =>
+                g.SetType(XLSparklineType.Stacked));
+
+            ws3.Cell("A2").Value = "Stacked, Colorful 1, All markers, SameForAll scale";
+            ws3.Cell("A5").Value = "Stacked, Colorful 2, First+Last+High+Low, Automatic scale";
+            ws3.Cell("A8").Value = "Stacked, Colorful 3, Markers+Negative, Custom scale";
+            ws3.Cell("A11").Value = "Stacked, Colorful 1, Date range";
+            ws3.Cell("A14").Value = "Stacked, Colorful 4, Line weight=2, Right to left";
+            ws3.Cell("A17").Value = "Stacked, Colorful 3, Different ranges";
+
```
Comment (igitur) [inline ClosedXML_Examples/Sparklines/SampleSparklines.cs]: Please autofit each worksheet's columns.

File: ClosedXML_Tests/Excel/Sparklines/SparklinesTests.cs [hunk 6]
```diff
@@ -0,0 +1,1075 @@
...
+
+namespace ClosedXML_Tests.Excel.Sparklines
+{
+    [TestFixture]
+    public class SparklinesTests
+    {
+        #region Add sparklines
+
+        [Test]
+        public void CanotCreateSparklineGroupsWithoutWorksheet()
```
Comment (igitur) [inline ClosedXML_Tests/Excel/Sparklines/SparklinesTests.cs]: `Canot` -> `Cannot`

File: ClosedXML_Tests/Excel/Sparklines/SparklinesTests.cs [hunk 7]
```diff
@@ -0,0 +1,1075 @@
...
+        {
+            TestDelegate action = () => new XLSparklineGroups(null);
+            Assert.Throws<ArgumentNullException>(action);
+        }
+
+        [Test]
+        public void CanotCreateSparklineGroupWithoutWorksheet()
```
Comment (igitur) [inline ClosedXML_Tests/Excel/Sparklines/SparklinesTests.cs]: `Canot` -> `Cannot`

File: ClosedXML_Tests/Excel/Sparklines/SparklinesTests.cs [hunk 8]
```diff
@@ -0,0 +1,1075 @@
+﻿using ClosedXML.Excel;
```
Comment (igitur) [inline ClosedXML_Tests/Excel/Sparklines/SparklinesTests.cs]: Do a CodeMaid organize on all *new* files and add the `// Keep ...` header, please.

Comment (igitur) [reply ClosedXML_Tests/Excel/Sparklines/SparklinesTests.cs]: This class is insanely thorough!

**Comment by Pankraty** [comment]
Done

**Comment by igitur** [comment]
The last 3 commits (*not* e6e3f198782711d9714a53024b79cc37036442b1) can be squashed, right?

**Comment by Pankraty** [comment]
Yes. Give me a couple of minutes

**Comment by Pankraty** [comment]
Done
